runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -125,40 +152,42 @@ private void invokeProcessElement(WindowedValue<InputT> elem) {
   public void finishBundle() {
     // This can contain user code. Wrap it in case it throws an exception.
     try {
-      fn.finishBundle(context);
+      invoker.invokeFinishBundle(context);
     } catch (Throwable t) {
       // Exception in user code.
       throw wrapUserCodeException(t);
     }
   }
 
-  /**
-   * Returns a new {@link OldDoFn.ProcessContext} for the given element.
-   */
-  private OldDoFn<InputT, OutputT>.ProcessContext createProcessContext(
-      WindowedValue<InputT> elem) {
+  /** Returns a new {@link DoFn.ProcessContext} for the given element. */
+  private DoFn<InputT, OutputT>.ProcessContext createProcessContext(WindowedValue<InputT> elem) {
     return new DoFnProcessContext<InputT, OutputT>(fn, context, elem);
   }
 
+  private DoFn.ExtraContextFactory<InputT, OutputT> createExtraContextFactory(
+      WindowedValue<InputT> elem) {
+    return new DoFnExtraContextFactory<InputT, OutputT>(elem.getWindows(), elem.getPane());
+  }
+
   private RuntimeException wrapUserCodeException(Throwable t) {
     throw UserCodeException.wrapIf(!isSystemDoFn(), t);
   }
 
   private boolean isSystemDoFn() {
-    return fn.getClass().isAnnotationPresent(SystemDoFnInternal.class);
+    return invoker.getClass().isAnnotationPresent(SystemDoFnInternal.class);
   }
 
   /**
-   * A concrete implementation of {@code OldDoFn.Context} used for running a {@link OldDoFn}.
+   * A concrete implementation of {@code DoFn.Context} used for running a {@link DoFn}.
    *
-   * @param <InputT> the type of the {@link OldDoFn} (main) input elements
-   * @param <OutputT> the type of the {@link OldDoFn} (main) output elements
+   * @param <InputT> the type of the {@link DoFn} (main) input elements
+   * @param <OutputT> the type of the {@link DoFn} (main) output elements
    */
-  private static class DoFnContext<InputT, OutputT> extends OldDoFn<InputT, OutputT>.Context {
+  private static class DoFnContext<InputT, OutputT> extends DoFn<InputT, OutputT>.Context {
```

Comment:
We should consider making DoFnContext mutable and reusing it across elements, or leaving it immutable but making it unnecessary.